src/RFC-0313_VNRegistration.md

```diff
@@ -0,0 +1,343 @@
+
+Additionally, we introduce a new [block header] field `validator_node_mr` that contains a [Merkle root] committing to all validator 
+`Vn_Shard_Key`s in the current epoch.
+The `validator_node_mr` needs to be recalculated at every `EpochSize` blocks to account for departing and arriving nodes.
+The `validator_node_mr` MUST remain unchanged for blocks between epochs, that is, blocks that are not multiples of `EpochSize`.
+
+> A validator generates a [Merkle proof] that proves its `VN_Shard_Key` is included in the validator set for any given epoch. This proof is provided in [Quorum Certificate]s.
+
+The `validator_node_mr` is calculated for each block as follows:
+
+```rust,ignore
+if block_height % EpochSize == 0 {
+    let vn_set = get_vn_set(chain_db, consensus_constants, current_epoch);
+    calculate_validate_node_mr(vn_set)
+} else {
+    let previous_block_header = get_previous_block_header(chain_db, block_height);
+    previous_block_header.validator_node_mr
+}
+```
+
+## Epoch transitions
+
+The [DAN] BFT consensus protocol relies on a shared and consistent "source of truth" from the [base layer] chain that defines
+the current epoch and validator set as well as [templates](./RFCD-0306_DANTemplateRegistration.md).
+
+As briefly mentioned in the [overview](#overview), any PoW [base layer] chain is prone to [reorg]s. The question arises, how do we 
+achieve a shared, consistent view of the chain when the data can disappear from underneath you?
+
+We define a `VNConfirmationPeriod` as is a network-wide constant that specifies the number confirmations (blocks) that a block must have 
+before a validator will recognise it as final. The chosen value for `VNConfirmationPeriod` must be large enough to make reorgs beyond that 
+point practically impossible. Validators simply ignore [base layer] [reorg]s with a depth _less than_ `VNConfirmationPeriod` deep as the 
+data they have extracted is still valid. This means that the _point of finality_ is not always `VNConfirmationPeriod` blocks away from the
+tip and is non-decreasing/monotonic which effectively negates [reorg] "noise" from the chain tip. 
+
+However, this does not address [base layer] latency delays where a single huge block or multi-block reorgs may take seconds to be received 
+and processed by all [base node]s. Moreover, the validators may poll the base layer only every few seconds, further increasing the latency for 
+validators to become aware of the state. This means a single strict validator epoch change-over point will almost always cause liveness failures at 
+and after the epoch transition.
+
+To address this, we define a `VNEpochGracePeriod` where both the previous and current epoch are accepted. This value, in blocks, must allow enough
+time for all validators to become aware of the [base layer] state for the epoch.
+
+To illustrate, consider the following view of a [base layer] chain. We mark 3 views of the chain.
+
+```text
```

Comment:
We need to take a look here wat happens with leader changes, on point b, we need to make a rule surrounding what leader to listen to, we might have two leaders in E11 and E12, both honest. 

which 2 do the nodes listen to? Their local epoch?

Reply:
As long as they only vote for one message out of the two leaders, it is fine